ipyparallel/controller/hub.py

```diff
@@ -833,15 +833,26 @@ def mia_task_request(self, idents, msg):
 
     #--------------------- IOPub Traffic ------------------------------
 
-    def save_iopub_message(self, topics, msg):
-        """save an iopub message into the db"""
-        # print (topics)
+    def monitor_iopub_message(self, topics, msg):
+        '''intercept iopub traffic so events can be acted upon'''
```

Comment:
The shutdown_reply should be safely treated the same as the engine being totally gone, because it will not process any more outstanding requests after this point (this would be a lot more complicated if you tried to unregister when you see the request go by).